src/main/java/gyro/google/kms/CryptoKeyResource.java

```diff
@@ -0,0 +1,382 @@
+            setCryptoKeyVersionTemplate(cryptoKeyVersionTemplate);
+        }
+
+        if (model.hasPrimary()) {
+            setPrimaryKeyVersionId(model.getPrimary().getName().split("/")[9]);
+        }
+    }
+
+    @Override
+    protected boolean doRefresh() throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        CryptoKey cryptoKey = client.getCryptoKey(getId());
+
+        if (cryptoKey == null) {
+            return false;
+        }
+
+        copyFrom(cryptoKey);
+        refreshVersions(client);
+
+        client.shutdownNow();
+
+        return true;
+    }
+
+    @Override
+    protected void doCreate(GyroUI ui, State state) throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        CryptoKey.Builder builder = CryptoKey.newBuilder()
+            .putAllLabels(getLabels())
+            .setPurpose(getPurpose())
+            .setVersionTemplate(getCryptoKeyVersionTemplate().toCryptoKeyVersionTemplate());
+
+        if (getRotationPeriod() != null) {
+            builder = builder.setRotationPeriod(Duration.newBuilder().setSeconds(getRotationPeriod() * 86400).build());
+        }
+
+        if (getNextRotationDate() != null) {
+            SimpleDateFormat format = new SimpleDateFormat("MM/dd/yyyy");
+            builder = builder.setNextRotationTime(Timestamp.newBuilder()
+                .setSeconds(format.parse(getNextRotationDate()).getTime() / 1000)
+                .build());
+        }
+
+        String parent = KeyRingName.format(
+            getProjectId(),
+            getKeyRing().getLocationFromId(),
+            getKeyRing().getNameFromId());
+
+        CryptoKey response = client.createCryptoKey(parent, getName(), builder.build());
+
+        if (getPrimaryKeyVersionId() != null) {
+            String cryptoKeyPathName = CryptoKeyPathName.format(
+                getProjectId(),
+                getKeyRing().getLocationFromId(),
+                getKeyRing().getNameFromId(),
+                getName());
+
+            client.updateCryptoKeyPrimaryVersion(cryptoKeyPathName, getPrimaryKeyVersionId());
+        }
+
+        copyFrom(response);
+
+        refreshVersions(client);
+
+        client.shutdownNow();
+    }
+
+    @Override
+    protected void doUpdate(
+        GyroUI ui, State state, Resource current, Set<String> changedFieldNames) throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        FieldMask.Builder fieldMaskBuilder = FieldMask.newBuilder();
+        CryptoKey.Builder cryptoKeyBuilder = CryptoKey.newBuilder().setName(getId());
+
+        if (changedFieldNames.contains("rotation-period")) {
+            cryptoKeyBuilder.setRotationPeriod(Duration.newBuilder().setSeconds(getRotationPeriod() * 86400).build());
+            fieldMaskBuilder.addPaths("rotation_period");
+        }
```

Comment:
NPE if `getRotationPeriod ` is set to null.

Reply:
This field can't be changed to null. It's a required based on the `purpose` which is not updatable.

Reply:
Is there a validation somewhere for that ?

Reply:
Yep, the `validate()` method does it.